core/market/src/negotiation/common.rs

```diff
@@ -576,8 +599,33 @@ pub fn expect_state(
     })?
 }
 
-fn verify_reason(reason: Option<&String>) -> Result<(), ReasonError> {
-    Ok(if let Some(s) = reason {
-        serde_json::from_str::<Reason>(s)?;
-    })
+fn get_reason_code(reason: &Option<Reason>, key: &str) -> Option<String> {
+    reason
+        .as_ref()
+        .map(|reason| {
+            reason
+                .extra
+                .get(key)
+                .map(|json| json.as_str().map(|code| code.to_string()))
+        })
+        .flatten()
+        .flatten()
+}
+
+/// This function extract from Reason additional information about termination reason
+/// and increments metric counter. Note that Reason isn't required to have any fields
+/// despite 'message'.
+pub fn terminate_reason_metric(reason: &Option<Reason>, owner: OwnerType) {
+    let p_code = get_reason_code(reason, "golem.provider.code");
+    let r_code = get_reason_code(reason, "golem.requestor.code");
+
+    let reason_code = r_code.xor(p_code).unwrap_or("None".to_string());
```

Comment:
to be consistent with ya-provider
```suggestion
    let reason_code = r_code.xor(p_code).unwrap_or("NotSpecified".to_string());
```